shared/src/main/java/io/crate/concurrent/CompletableFutures.java

```diff
@@ -57,4 +65,345 @@ private CompletableFutures() {
         }
     }
 
+    public static <T> CompletableFuture<T> runBeforeCompleteExceptional(CompletableFuture<T> delegate,
+                                                                        Consumer<Throwable> beforeExceptional) {
+        return new ExceptionalInterceptingFuture<>(delegate, beforeExceptional);
+    }
+
+    @SuppressForbidden
+    private static class ExceptionalInterceptingFuture<T> extends CompletableFuture<T> {
+
+        private final CompletableFuture<T> delegate;
+        private final Consumer<Throwable> beforeExceptional;
+
+        private ExceptionalInterceptingFuture(CompletableFuture<T> delegate, Consumer<Throwable> beforeExceptional) {
+            this.delegate = delegate;
+            this.beforeExceptional = beforeExceptional;
+        }
+
+        @Override
+        public boolean completeExceptionally(Throwable ex) {
+            beforeExceptional.accept(ex);
+            return delegate.completeExceptionally(ex);
```

Comment:
I thought the plan was to intercept `cancel` instead, and not really forward it, but that due to the intercepted `cancel` handling the operation encapsulated would then eventually finish (with a failure) by itself.


```
future
  .cancel()
      ------------------> OrderedDocCollector
                             raiseError
                                |
     <--------------------------+
 Future
  completed with error
```


Instead of


```
future
  .completeExceptionally(err)
      inercept
      ------------------> OrderedDocCollector
                             raiseError
     <--------------------------+
  delegate.completeExcep.       |
                                |
     <--------------------------+
 Future
  completed with error
```

Is this due some `cancel` specific implementations, or because it doesn't allow to propagat a cause (Throwable)?

Reply:
It's because `cancel` does not allow any cause propagation and always results in a `CancellationException` set on the future.
`cancel` and `completeExceptionally` implementations of a `CompletableFuture` are almost similary, only the exception is different.

Reply:
To me they also imply a semantic difference. 

Either way, whether it is cancel or completeExceptionally, I think the delegate shouldn't be called in addition.

```
--- a/shared/src/main/java/io/crate/concurrent/CompletableFutures.java
+++ b/shared/src/main/java/io/crate/concurrent/CompletableFutures.java
@@ -84,7 +84,7 @@ public final class CompletableFutures {
         @Override
         public boolean completeExceptionally(Throwable ex) {
             beforeExceptional.accept(ex);
-            return delegate.completeExceptionally(ex);
+            return true;
         }

         @Override
```

Because `beforeExceptional.accept(ex); ` should result in the future to be completed from *inside*  by the other thread. We don't want to complete the future from both inside and outside.

Reply:
> We don't want to complete the future from both inside and outside.

Hm ok this was my initial critic on that approach but seems like we had some misunderstanding and meant the same.

Reply:
On the other hand, by raising an exception inside the collect thread, the future *will*  be set from the inside as e.g. `CompletableFuture.applyAsync` will catch it and completes exceptionally. So I think it should never be set from outside by never calling the delegate.

Reply:
>  So I think it should never be set from outside by never calling the delegate.

That's what I'm saying. See my proposed diff and the ascii art further above.

I'm inclined to go a step further and change some of the methods/interfaces to return a  `CompletionStage`, so that attaching events is still possible but to hide the `complete..` methods. We should never allow to complete some action from outside.

To cancel it we'd have to provide a wrapper for the `CompletionStage` that also implements `Killable` or something like that.

Reply:
> We should never allow to complete some action from outside.

Ok this was my point from the very beginning, seem like I had issues to communicate this correctly.